proposals/0672-declarative-xcode-projects.md

```diff
@@ -0,0 +1,259 @@
+
+The suggested solution is to adopt [Tuist.io](https://tuist.io) in the App template. This allows to describe the project in Swift, the default language for Apple development.
+
+This tool offers some immediate benefits, depending on the user's category:
+- For the React Native team, less maintenance cost. We only need to maintain a file that is ~100 LOC rather than ~1200 LOC.
+- For React Native users that are using an app created from the CLI:
+    - fewer code in user space (again, 100 LOC rather than 1200).
+    - easier way to spot difference when upgrading their iOS project.
+- For React Native users of brownfield applications:
+    - a clearer description of the public APIs of the build system that are needed to integrate React Native in their Native app.
+- For framework developers:
+    - a clearer description of the public APIs of the build system that are needed to build a React Native app, and a clearer starting point for the framework.
+- For everyone: adding/renaming iOS files in React Native project does not have to happen within Xcode.
+
+Moreover, the project description file is executable Swift code that runs when creating the project. We can use it to run some scripts ahead of time. This provide more flexibility than a pure markup language.
+
+## Basic example
+
+You can have a look at it in action in this PR:
+
+- https://github.com/facebook/react-native/pull/37952
+
+The idea is to describe the project with a declarative approach using Swift. Tuist allows you to load the `Project.swift` file in an instance of Xcode throught the `tuist edit` command. This allows to leverage code completion, type checking and syntax highlight to ensure the correctness of the project. The project description can also be built by Xcode, to verify that it will generate something valid.
+
+### What does it means for the final users
+
+We can separate the usage of the tool in different categories, based on when the tool comes into play:
+
+#### 1. For a user creating a new app from the Template.
+
+Nothing changes.
+The user will run:
+
+```bash
+npx react-native init <AppName>
+```
+
+This will:
+- download the template  - already doing that
+- run `yarn` - already doing that
+- eventually installing Tuist<sup>*</sup> (more on this later) - _new step_
+- run `tuist generate` to create the project - _new step_
+- run `bundler install` - already doing that
+- run `pod install` - already doing that
+
+From the user perspective, the two new steps will be transparent.
+
+#### 2. For a user upgrading a CLI app from a version without Tuist to a version with Tuist (e.g.: from 0.72 to 0.73)
+
+When performing the upgrade, the upgrade-helper will show the following steps:
+
+- Remove the `xcodeproj` file, removing a file that is 1200 LOC
```

Comment:
This may just be a wording issue, but I would be a little careful deleting an xcodeproj file without making it very clear to the user what the ramifications of this is.

While I agree that xcodeproj files are not a great way to manage iOS projects, they are what users have been using for a long time. At this point, they probably don't even remember what custom settings they have applied on top of the RN base project. When that file is removed, that information will be lost, and they'll be left with build failures and confusion.

In addition, Tuist will be a little bit more strict about where files are located. Again, I see this as a benefit of the tool, but existing users have definitely thrown there files around in random places, and regenerating their project won't just result in build failures, but also possibly runtime issues having to do with where they put their resources (images, localization files, etc).

Reply:
I see the point that regenerating the project if there are custom settings could be tricky.

However, the proposal aims to remove the xocdeproj from the template (so new pojects won't need it). Users that upgrades won't see their Xcodeproj removed one day from the other - the CLI will not remove it.

[Tuist offers some guides](https://docs.tuist.io/guides/adopting-tuist#extract-your-build-settings-into-xcconfig-files) on how to migrate away from the Xcodeproj. We may add links to those guides to help users migrate away from their xcodeproj.

The ultimate goal for projects would be not to have those files as they are sources of conflicts and hard to maintain. I agree that it is a delicate process, but we will all be in a better place if we manage to achieve that.

As first step, though, we may keep shipping the `Xcodeproj` files alongside `Tuist`, and then evolve the solution starting from there. How does that sound?